tensornetwork/backends/jax/jitted_functions.py

```diff
@@ -593,28 +601,341 @@ def gmres_residual(A_mv: Callable, A_args: Sequence, b: jax.ShapedArray,
+                                           tol,
+                                           b_norm)
+    x = gmres_update(k, V, R, beta_vec, x0)
+    return x, done
+
+
+  @jax.jit
+  def gmres_scan(gmres_carry, gmres_xs):
+    """
+    This function wraps the main GMRES loop for compatibility with
+    jax.lax.scan. If the error estimate err has dropped beneath the
+    convergence threshold, no work is done at this iteration, so that the
+    loop effectively terminates. Otherwise we call the main loop
+    `gmres_work'.
+
+    Args
+    ----
+    gmres_carry: The `carry` argument to jax.lax.scan.
+     A nested tuple (gmres_variables, gmres_constants), themselves containing:
+     gmres_variables = (final_k, V, R, beta_vec, err, givens):
+      -final_k: The integer loop index. It is incremented until err has
+                converged
+                beneath tol and then held constant.
+      -V      : An (n, n_kry + 1) array that will be used to store Krylov
+                vectors during the internal Arnoldi process. It should be
+                initialized as zeroes apart from V[0, :], the initial normalized
+                Krylov vector.
+      -R      : An (n_kry + 1, n_kry) array that will be used to store the R
+                factor in the QR decomposition of the matrix of overlaps
+                obtained during the Arnoldi process. It should be initialized
+                as zeroes.
+      -beta_vec: An (n_kry + 1) array that will be used to store the residual
+                 between the solution vector and its projection into the Krylov
+                 space. It should be initialized as zeroes apart from
+                 beta_vec[0], the magnitude of the initial Krylov vector before
+                 it was normalized.
+      -err    : A real float that estimates the relativie magnitude of this
+                residual at each iteration (the magnitude divided by that of
+                the b in Ax = b).
+      -givens : A (2, n_kry + 1) array that will be used to store the Givens
+                rotation factors used to compute R (the upper triangular factor
+                in the QR factorization of H) from H (the matrix of overlaps
+                obtained during the Arnoldi process). It should be initialized
+                as zeroes.
+    gmres_constants:
+      -tol  : The scan effectively terminates when err < tol by making
+              subsequent iterations null ops.
+      -A_mv : The linear operator representing A in A x = b.
+      -A_args: A_mv is called as A_mv(*A_args, v).
+      -b_norm: Norm of the b in A x = b.
+
+    Returns
+    -------
+    The same entries as in gmres_carry, updated as described above.
+    """
+    _ = gmres_xs
+    gmres_variables, gmres_constants = gmres_carry
+    err = gmres_variables[4]
+    tol = gmres_constants[0]
+    gmres_carry = jax.lax.cond(err < tol,
+                               gmres_carry, null_op,
+                               gmres_carry, gmres_work)
```

Comment:
So this may not work the way you expect. In XLA (especially on TPU), it'll actually evaluate BOTH methods everytime and just return the correct one based on the conditional.

Reply:
Also, the API changed for cond to make it simpler. https://jax.readthedocs.io/en/latest/_autosummary/jax.lax.cond.html?highlight=cond#jax.lax.cond

Can you change this to 

```python
jax.lax.cond(err < tol, null_op, gmres_work, gmres_carry)
```

Reply:
I see. That's a significant enough problem that I'll probably need to rewrite the whole thing, actually.